### DISCUSSION TITLE
openmc built with dagmc?

### DISCUSSION THREAD

**Discussion by veeshy** [neams-th-coe/cardinal#534] [General]
Looking to run a cad based geometry with a mesh tally over it, I installed `dagmc` via conda and then added `-DOPENMC_USE_DAGMC=on \` to the `openmc.mk` file, it fails with:

```
[  4%] Building CXX object CMakeFiles/libopenmc.dir/src/dagmc.cpp.o
/Users/vishalpatel/mambaforge3/envs/moose/bin/mpicxx  -std=gnu++17 -DDAGMC -DFMT_LOCALE -DGIT_SHA1=\"f5bd55debd7c16560f5376cb4696b7c10abc3084\" -DGSL_THROW_ON_CONTRACT_VIOLATION -DLIBMESH -DOPENMC_MPI -DPHDF5 -DUSE_LIBPNG -Dgsl_CONFIG_ALLOWS_NONSTRICT_SPAN_COMPARISON=1 -Dgsl_CONFIG_DEFAULTS_VERSION=1 -Dlibopenmc_EXPORTS -I/Users/vishalpatel/projects/cardinal/contrib/openmc/include -I/Users/vishalpatel/projects/cardinal/build/openmc/include -isystem /Users/vishalpatel/projects/cardinal/install/include -isystem /Users/vishalpatel/mambaforge3/envs/moose -isystem /Users/vishalpatel/mambaforge3/envs/moose/include/eigen3 -isystem /Users/vishalpatel/mambaforge3/envs/moose/libmesh/include -isystem /Users/vishalpatel/mambaforge3/envs/moose/libmesh-vtk/include/vtk-9.1 -O2 -felide-constructors -Qunused-arguments -Wunused-parameter -Wunused -fopenmp -DMETHOD=opt -Werror=return-type -Werror=reorder -Woverlength-strings -DENABLE_OPENMC_COUPLING -DNEKRS_VERSION=21 -DNEKRS_SUBVERSION=1 -DGITCOMMITHASH=\"51d5bf5f2042e231d1770400c160d5623b19b4c8\" -O3 -DNDEBUG -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -fPIC -DH5Oget_info_by_idx_vers=1 -DH5O_info_t_vers=1 -fopenmp=libomp -DNDEBUG -D_THREAD_SAFE -pthread -std=c++17 -MD -MT CMakeFiles/libopenmc.dir/src/dagmc.cpp.o -MF CMakeFiles/libopenmc.dir/src/dagmc.cpp.o.d -o CMakeFiles/libopenmc.dir/src/dagmc.cpp.o -c /Users/vishalpatel/projects/cardinal/contrib/openmc/src/dagmc.cpp
/Users/vishalpatel/mambaforge3/envs/moose/include/pyne.h:6104:23: error: 'read_e_bounds' has C-linkage specified, but returns user-defined type 'std::vector<double>' which is incompatible with C [-Werror,-Wreturn-type-c-linkage]
  std::vector<double> read_e_bounds(std::string e_bounds_file);
```

I imagine it's some sort of build env issue considering openmc docker image can build with libmesh+dagmc. Anyone have any luck with this?

**Comment by aprilnovak** [comment]
Hi @veeshy - is the end goal to couple a DAGMC OpenMC model to MOOSE? If so, Cardinal at the moment only supports data transfers to/from cell-based OpenMC geometries. 

That said, when we build Cardinal we build two executables: a Cardinal executable (`cardinal-opt`) and a "standalone" OpenMC executable (`openmc`). I _think_ you should be able to configure Cardinal with DAGMC turned on in the OpenMC build in order to run DAGMC cases with the `openmc` executable (but you wouldn't be able to _couple_ that DAGMC model to MOOSE). I could see someone doing this if they wanted to avoid having multiple OpenMC builds on their system, but otherwise there's limited utility to this.

You may already be aware of it, but there is an open-source coupling of DAGMC-based OpenMC models to MOOSE, called [Aurora](https://github.com/aurora-multiphysics/aurora) developed by the UKAEA. 

tagging @pshriwise in case he has anything to add

**Reply by veeshy** [reply]
Ah gotcha, thanks for that clarification. At this stage I just wanted one way coupling openmc-dagmc geometry tallying on a libmesh mesh to use as a heat source in moose. I hadn't seen Aurora actually so thanks for pointing that out.

**Reply by aprilnovak** [reply]
If you're only interested in one-way coupling, we might be able to support that in Cardinal with fairly little effort... the functions we use to write from a mesh tally to MOOSE are agnostic of the underlying OpenMC geometry. I think we'd just need to build Cardinal with DAGMC turned on, and then move the code that copies an OpenMC mesh tally to MOOSE into its own AuxKernel. 

@veeshy whatever you decide on, I definitely help with Cardinal, and can do my best to help with Aurora too. From a user perspective, you may find Cardinal more approachable since it has more documentation and is designed a bit differently (Aurora currently has a few restrictions on how it's used - can elaborate more if you'd like).

**Reply by pshriwise** [reply]
Hi @veeshy! 

I just wanted to clarify a couple points on capability.

DAGMC models in OpenMC are represented as cells just as they would be with a CSG model, so the cell-based coupling provided by Cardinal should work "out of the box" with a DAGMC model. I've done this before for one of the Cardinal pincell tests I believe. @aprilnovak, has it right about adding DAGMC support to the openmc library that is compiled as part of the Cardinal build -- that might take some extra work.

As for transfers using a libmesh mesh, this can be done currently using an unstructured mesh tally where the mesh is superimposed on the geometry. However, we can't yet treat the libmesh mesh as geometry where each element has it's own temperature, density, and material, so that information would only be fed back into the OpenMC model for the cells. 

I hope that made sense. Please let me know if you have any other questions.

**Reply by veeshy** [reply]
yup made sense @pshriwise, thanks! That was my mental model going in too so that's nice. 

@aprilnovak I took a peak at aurora and saw some of the restrictions, no csg+dagmc in one file and I think the FEM mesh is passed to openmc for tallying rather than being able to use a separate mesh and use transfers make me lean towards cardinal, but aurora looks dang cool still.

As for building with cardinal's openmc with dagmc, is the play to build dagmc from source too?

@aprilnovak I didn't quote follow on the openmc mesh tally copier to an auxkernel part, is this a different step than providing `solid_blocks` to the `OpenMCCellAverageProblem` problem to get tallies on the inputted mesh?

**Reply by aprilnovak** [reply]
Thanks @pshriwise for the input!

I was thinking too narrowly before - Patrick is correct, we should be able to run using DAGMC as-is within Cardinal, as long as the tally mesh is just superimposed on the CAD domain. In other words, there's two meshes - a DAGMC surface mesh that OpenMC tracks particles on, and a libMesh mesh used for tallying (which is overlaid on the geometry). 

Yes, we'll need to build DAGMC from source and link it to OpenMC when building Cardinal. I think this should be straightforward (I'm not 100% sure on the error you saw when attempting this, maybe it goes away if we do from source).

Let me try to set up a minimum example.

**Reply by pshriwise** [reply]
> As for building with cardinal's openmc with dagmc, is the play to build dagmc from source too?

That's what I'd recommend. There's a mesh library, MOAB, that will be an added dependency to your stack too.
And, optionally, there are a couple more optional packages for DAGMC that will enhance its ray tracing performance.

I think a good first step would be to compile OpenMC on its own with DAGMC enabled to make sure you have that worked out and then move on to adding DAGMC to the OpenMC that's built with Cardinal.

**Reply by pshriwise** [reply]
One caveat: MOAB needs to be built with HDF5 support enabled. HDF5 is already a requirement for OpenMC, so hopefully that is just a matter of setting the `ENABLE_HDF5` CMake option correctly.

**Reply by aprilnovak** [reply]
@veeshy I will put together the Makefile/etc. changes needed to build DAGMC with Cardinal. 

In the meantime, if you want to get started with Cardinal, all of the tutorials/etc. should still apply to your use case (with a DAGMC geometry in place of a cell geometry). I'll hopefully have something soon.

**Reply by veeshy** [reply]
Awesome! Ahead of you there, I didn't do my research on what openmc was built with until it was too late though ;), I got a cell based version going (well kind of, my geometry needs the cad to be correct)

**Reply by aprilnovak** [reply]
I've got a simple MWE working that couples a DAGMC OpenMC model to MOOSE. I took the [pincell DAGMC example from OpenMC](https://github.com/openmc-dev/openmc-notebooks/blob/main/cad-based-geometry.ipynb) and threw together a quick libMesh cube hex mesh. Here's the image of the OpenMC heat source computed on a DAGMC model, mapped to a libMesh mesh (far right):

![Screen Shot 2022-12-14 at 6 48 03 PM](https://user-images.githubusercontent.com/17039662/207760876-168b8303-2fcf-4c71-b16e-60268e4b607e.png)

In order to build DAGMC within Cardinal, I need a [small fix to go into Pyne](https://github.com/pyne/pyne/pull/1467). After that's merged, I'll put this formally into Cardinal to support it - should be ready in a few days?

**Reply by aprilnovak** [reply]
FWIW, I was able to reproduce @veeshy's build error with MOOSE's conda environment. Seems like a minor fix arising from error checks that MOOSE enforces, but Pyne doesn't: https://github.com/pyne/pyne/pull/1469
